### PR TITLE
fix!: create a new CML scope for every call of BuildTx in selection constraints

### PR DESCRIPTION
**Pull request by mirceahasegan**
# Context
Calls computeMinimumCoinQuantity 4 times for every tx change for each asset. If tx is changed too fast it errors with undefined and then RangeError: offset is out of bounds.

# Proposed Solution
Fix some memory leaks when calling CML.

### CODE REVIEW

File: packages/core/src/Cardano/util/address.ts [hunk 1]
```diff
@@ -4,7 +4,13 @@ import { CML } from '../../CML/CML';
 /**
  * Validate input as a Cardano Address from all Cardano eras and networks
  */
-export const isAddress = (input: string): boolean => CML.Address.is_valid(input);
+export const isAddress = (input: string): boolean => {
+  try {
+    return CML.Address.is_valid(input);
```
[the remark this answers is not in the document]
Comment (mirceahasegan) [reply packages/core/src/Cardano/util/address.ts:9]: The throw happens when the memory has become invalid (unexpected null pointer, etc..).

[the remark this answers is not in the document]
Comment (mirceahasegan) [reply packages/core/src/Cardano/util/address.ts:9]: Yes, it became unusable due to the corrupted memory. Maybe we shouldn't encapsulate it if that is the only scenario where it throws. I'll let @mkazlauskas chip in, as there may be more scenarios 😄

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/core/src/Cardano/util/address.ts:9]: Agree, let's drop the commit

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
If we can't reproduce in unit tests, maybe let's add a web-extension test in e2e package

File: packages/core/src/Cardano/util/address.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mirceahasegan) [reply packages/core/src/Cardano/util/address.ts]: dropped the commit and force pushed

File: packages/input-selection/src/selectionConstraints.ts [hunk 2]
```diff
@@ -55,10 +55,14 @@ export const computeMinimumCoinQuantity =
...
     usingAutoFree((scope) =>
       BigInt(
-        CML.min_ada_required(
-          coreToCml.txOut(scope, output),
-          scope.manage(CML.BigNum.from_str(coinsPerUtxoByte.toString()))
-        ).to_str()
+        scope
+          .manage(
+            CML.min_ada_required(
+              scope.manage(coreToCml.txOut(scope, output)),
```
Comment (rhyslbw) [inline packages/input-selection/src/selectionConstraints.ts:61]: We've had the convention of managing any objects within the `coreTo*` translators. Can it not be done using the scope passed in?

File: packages/input-selection/src/selectionConstraints.ts [hunk 3]
```diff
@@ -70,7 +74,7 @@ export const tokenBundleSizeExceedsLimit =
     }
     return usingAutoFree((scope) => {
       const value = scope.manage(CML.Value.new(cmlUtil.maxBigNum));
-      value.set_multiasset(coreToCml.tokenMap(scope, tokenBundle));
+      value.set_multiasset(scope.manage(coreToCml.tokenMap(scope, tokenBundle)));
```
Comment (rhyslbw) [inline packages/input-selection/src/selectionConstraints.ts:77]: Here too

File: packages/core/src/CML/coreToCml/coreToCml.ts [hunk 4]
```diff
@@ -90,11 +90,12 @@ export const tokenMap = (scope: ManagedFreeableScope, map: Cardano.TokenMap) =>
 
 export const value = (scope: ManagedFreeableScope, { coins, assets }: Cardano.Value): Value => {
   const result = scope.manage(Value.new(scope.manage(BigNum.from_str(coins.toString()))));
+
   if (!assets) {
     return result;
   }
   if (assets.size > 0) {
-    result.set_multiasset(tokenMap(scope, assets));
+    result.set_multiasset(scope.manage(tokenMap(scope, assets)));
```
Comment (rhyslbw) [inline packages/core/src/CML/coreToCml/coreToCml.ts:98]: I'm a bit confused about why this changes the behavior when the returned object is already being managed here: https://github.com/input-output-hk/cardano-js-sdk/blob/a29639715f47d7d5cab7d7aa667581bb1030620f/packages/core/src/CML/coreToCml/coreToCml.ts#L63

File: packages/input-selection/src/selectionConstraints.ts (hunk 2 above)
Comment (rhyslbw) [reply packages/input-selection/src/selectionConstraints.ts:61]: OK, I can see it's already managed internally here: https://github.com/input-output-hk/cardano-js-sdk/blob/a29639715f47d7d5cab7d7aa667581bb1030620f/packages/core/src/CML/coreToCml/coreToCml.ts#L112

:thinking:

Comment (mkazlauskas) [reply packages/input-selection/src/selectionConstraints.ts:61]: > We've had the convention of managing any objects within the coreTo* translators. Can it not be done using the scope passed in?

Good point and yes the `txOut` and `tokenMap` scope.manage in this PR seems redundant. Then the only functional change in this PR seems to be `scope.manage(CML.min_ada_required(`. @mirceahasegan could you please test to verify?

File: packages/input-selection/src/selectionConstraints.ts [hunk 5]
```diff
@@ -87,14 +90,15 @@ const getTxSize = (tx: CML.Transaction) => tx.to_bytes().length;
...
-    const tx = await buildTx(selectionSkeleton);
-    const txSize = getTxSize(tx);
-    if (txSize <= maxTxSize) {
-      return selectionSkeleton.inputs.size;
-    }
-    return selectionSkeleton.inputs.size - 1;
-  };
+  (selectionSkeleton) =>
+    usingAutoFree(async (scope) => {
+      const tx = scope.manage(await buildTx(selectionSkeleton, scope));
```
Comment (rhyslbw) [inline packages/input-selection/src/selectionConstraints.ts:95]:
```suggestion
      const tx = await buildTx(selectionSkeleton, scope);
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/input-selection/src/selectionConstraints.ts (hunk 5 above)
Comment (mirceahasegan) [reply packages/input-selection/src/selectionConstraints.ts]: [b5fcf86](https://github.com/input-output-hk/cardano-js-sdk/pull/534/commits/b5fcf86fb2f79e599e3731b79f3038023e43f679)

**Review by rhyslbw — DISMISSED** [review]

**Review by rhyslbw — APPROVED** [review]

**Review by mkazlauskas — DISMISSED** [review]
[BigNumu](https://github.com/input-output-hk/cardano-js-sdk/pull/534/commits/3b9ec98e06ea85bd2274258fae195dfeb9039406) :grin:

**Review by mkazlauskas — DISMISSED** [review]

**Review by rhyslbw — APPROVED** [review]

**Review by mkazlauskas — APPROVED** [review]
